File: algorithms/[216]组合总和 III/solution.py
# topics = ["回溯算法"]

from typing import List


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        res: List[List[int]] = []
        path: List[int] = []

        def backtrack(idx: int = 1) -> None:
            size, total = len(path), sum(path)
            if total == n and size == k:
                res.append(path[:])
                return
            if size == k:
                return
            for i in range(idx, 10):
                if total + i > n:
                    return
                path.append(i)
                backtrack(i + 1)
                path.pop()

        backtrack()

        return res

File: algorithms/[216]组合总和 III/solution.ts
// topics = ["回溯算法"]

/**
 * 回溯法搜索所有可行解模板
 * res = []
 * path = []
 *
 * def backtrack(未探索区域, res, path):
 *     if path 满足条件:
 *         res.add(path)  # 深度拷贝
 *         # return       # 如果不用继续搜索需要 return
 *     for 选择 in 未探索区域当前可能的选择:
 *         if 当前选择符合要求:
 *             path.add(当前选择)
 *             backtrack(新的未探索区域, res, path)
 *             path.pop()
 */

// 近似题 40.组合总和 II
function combinationSum3(k: number, n: number): number[][] {
  const res: number[][] = [];
  const path: number[] = [];

  const backtrack = (idx = 1): void => {
    const sum = path.reduce((acc, cur) => acc + cur, 0);
    const len = path.length;
    if (sum == n && len == k) {
      res.push([...path]);
      return;
    }
    if (len == k) {
      return;
    }
    for (let i = idx; i < 10; i += 1) {
      if (sum + i > n) {
        return;
      }
      path.push(i);
      backtrack(i + 1);
      path.pop();
    }
  };

  backtrack();

  return res;
}
